File: analysis/indicator_saturation_analysis.py
import logging
import multiprocessing
import sys
import pandas as pd
import argparse

from utilities import OUTPUT_DIR
#from study_definition import indicators_list

### Importing the local change_detection functions
# sys.path.append('bin')
from change_detection import functions as chg

indicators_list = ["a", "b", "c", "d", "e", "f", "g",
                   "i", "k", "ac", "me_no_fbc", "me_no_lft", "li", "am"]

if __name__ == '__main__':

    argv = sys.argv[1:]

    p = argparse.ArgumentParser(description="Run an indicator saturation analysis",
    allow_abbrev=True)

    p.add_argument('-indicator', type=str, help='the indicator id',
                   choices=indicators_list, required=True)
    p.add_argument('-numerator', type=str, help='the numerator column')
    p.add_argument('-denominator', type=str, help='the denominator column')

    args = p.parse_args()        

    if args.numerator == None:
        args.numerator = f"indicator_{args.indicator}_numerator"

    if args.denominator==None:
        args.denominator = f"indicator_{args.indicator}_denominator"

    multiprocessing.log_to_stderr()
    logger = multiprocessing.get_logger()
    logger.setLevel(logging.INFO)
        
    measures_file = f"{OUTPUT_DIR}/measure_indicator_{args.indicator}_rate.csv"

    print( f"Processing measure [{args.indicator}]: {measures_file}")

    measure_indicator = chg.ChangeDetection(f'indicator_saturation_{args.indicator}',
                                    verbose=True,
                                    code_variable="practice",
                                    numerator_variable=args.numerator,
                                    denominator_variable=args.denominator,
                                    date_variable="date",
                                    date_format="%Y-%m-%d",
                                    overwrite=True,
                                    draw_figures='yes',
                                    base_dir=OUTPUT_DIR,
                                    data_subdir='indicator_saturation',
                                    csv_name=measures_file)
    measure_indicator.run()

File: analysis/indicator_saturation_analysis.R
source("analysis/indicator_saturation_functions.R")

#####################################################################
### Parsing arguments ###############################################
#####################################################################

opt_parser = OptionParser(usage = "%prog [options] csv_file",
                          option_list=option_list )

opt = parse_args( opt_parser, positional_arguments = TRUE )

fatal_missing_args = c()

### Checking for missing required arguments/problematic inputs ######

if ( length( opt$args) == 0 ) {
    fatal_missing_args = c( fatal_missing_args,
                            "You must supply a file name" )
}

if (is.null(opt$options$indicator)){
    fatal_missing_args = c( fatal_missing_args,
                            "You must supply an indicator name" )
}

if ( opt$options$numcores < 1 ) {
    fatal_missing_args = c( fatal_missing_args,
                            "You must request at least 2 cores for processing" )
}

if ( length(fatal_missing_args) > 0 ) {
    print_help(opt_parser)
    messages = glue("- {fatal_missing_args}")
    stop(report_error( default_values, paste(messages,collapse='\n') ), call.=FALSE)
}

### Assuming numerators/denominators if not supplied ################

if (is.null( opt$options$numerator ) ) {
    opt$options$numerator = sprintf( "indicator_%s_numerator", opt$options$indicator )
}

if (is.null( opt$options$denominator ) ) {
    opt$options$denominator = sprintf( "indicator_%s_denominator", opt$options$indicator )
}

for ( this_direction in c( "up","down" ) ) {
    
    measure_indicator = ChangeDetection(
        name = glue('indicator_saturation_{this_direction}_{opt$options$indicator}'),
        code_variable = opt$options$code,
        numerator_variable = opt$options$numerator,
        denominator_variable = opt$options$denominator,
        date_variable = unlist(opt$options$date),
        date_format = opt$options$ymd,
        numcores = opt$options$numcores,
        indir = opt$options$indir,
        outdir = opt$options$outdir,
        direction = this_direction,
        overwrite = opt$options$overwrite,
        draw_figures = opt$options$figures,
        verbose = opt$options$verbose,
        test_number = opt$options$test,
        csv_name = opt$args[1]
    )
    
    run( measure_indicator )
    
}


### To test:
### Rscript indicator_saturation_analysis.R -I a -T date -C practice -v -i ../output -o ../output/indicator_saturation2 measure_indicator_a_rate.csv -f
